File: user-auth-service/src/main/java/com/damian/marketgrid/user/messaging/event/KafkaEvent.java
package com.damian.marketgrid.user.messaging.event;

import java.time.Instant;
import java.util.UUID;

public record KafkaEvent<T>(
        String eventId,
        Instant occurredAt,
        T payload
){
    public KafkaEvent(T payload) {
        this(UUID.randomUUID().toString(), Instant.now(), payload);
    }
}

File: ingest-service/src/main/kotlin/com/damian/marketgrid/messaging/event/KafkaEvent.kt
package com.damian.marketgrid.messaging.event

import java.time.Instant
import java.util.UUID

data class KafkaEvent<T>(
    val eventId: String,
    val occurredAt: Instant,
    val payload: T
) {
    companion object {
        fun <T> of(payload: T): KafkaEvent<T> =
            KafkaEvent(UUID.randomUUID().toString(), Instant.now(), payload)
    }
}
